Change from clients.openWindow to event.openWindow.

tommythorsen opened the pull request:
This change introduces a new method on the PaymentRequestEvent, called
openWindow(). It is similar to clients.openWindow, but since it has a
tighter relationship to the PaymentRequest, it should be more convenient
for implementors to provide an experience that fits the payment flow.

See https://github.com/w3c/webpayments-payment-apps-api/issues/97.

Commit:
Change from clients.openWindow to event.openWindow.

This change introduces a new method on the PaymentRequestEvent, called
openWindow(). It is similar to clients.openWindow, but since it has a
tighter relationship to the PaymentRequest, it should be more convenient
for implementors to provide an experience that fits the payment flow.

adrianhopebailie reviewed (approved):
LGTM, thanks @tommythorsen

tommythorsen commented:
I have updated the PR to sit on top of @ianbjacobs' restructuring. Does it still look ok?

marcoscaceres commented on index.html:
```diff
@@ -889,6 +889,7 @@
       [Exposed=ServiceWorker]
       interface PaymentRequestEvent : ExtendableEvent {
         readonly attribute PaymentAppRequest appRequest;
+        Promise<WindowClient> openWindow(DOMString url);
```
URL are USVStrings

tommythorsen replied:
From https://heycam.github.io/webidl/#idl-USVString:
> Specifications should only use USVString for APIs that perform text processing and need a string of Unicode scalar values to operate on. Most APIs that use strings should instead be using DOMString, which does not make any interpretations of the code units in the string. When in doubt, use DOMString.

I was in doubt, so I used DOMString. What calls for USVString in this situation?

marcoscaceres commented on index.html:
```diff
@@ -898,6 +899,11 @@
           This attribute contains the <a>payment app request</a> associated
           with this <a>payment request</a>.
         </dd>
+        <dt><code>openWindow</code> method</dt>
```
Please define methods in their own section.

tommythorsen replied:
I've done this, but now this section is inconsistent with the rest of the document. I can go through and give the other sections the same treatment, but that will be a separate PR.

marcoscaceres commented on index.html:
```diff
@@ -898,6 +899,11 @@
+        <dt><code>openWindow</code> method</dt>
+        <dd>
+          This method is used by the payment handler to show a window to
+          the user.
```
Should like to algorithm that describes steps

marcoscaceres commented on index.html:
```diff
@@ -964,9 +970,15 @@
-  <p class="issue">See <a href="https://github.com/w3c/webpayments-payment-apps-api/issues/97">issue 97</a> for discussion about opening window in a way that is consistent with payment flow and not confusing to the user. For example, there is consensus that in a desktop environment, a payment handler should <em>not</em> open a window in a new browser tab, as this is too dissociated from the check ...
-  
-  <p>User agents SHOULD display the origin of a running payment handler to the user.</p>
+  <p>A payment handler that requires visual display and user interaction, may call <code>PaymentRequestEvent.openWindow()</code>.
```
nit: `<code>PaymentRequestEvent.openWindow()</code>.` <a>openWindow()</a>`.

marcoscaceres commented on index.html:
```diff
@@ -964,9 +970,15 @@
+    Since user agents know that this method is connected to the payment request event, they SHOULD render the window in a way that is consistent with the flow and not confusing to the user.</p>
+
+  <p>The resulting window client is bound to the tab/window that initiated the <a>payment request</a>.
+    A single <a>payment handler</a> SHOULD NOT be allowed to open more than one client window using this method.
```
All this needs to be in algorithm.

tommythorsen replied:
I'm working on this. Aiming to update this PR with an algorithm for openWindow tomorrow.

marcoscaceres commented on index.html:
```diff
@@ -964,9 +970,15 @@
+
+  <p>The resulting window client is bound to the tab/window that initiated the <a>payment request</a>.
+    A single <a>payment handler</a> SHOULD NOT be allowed to open more than one client window using this method.
+    The exact behavior in the case where a payment handler calls <code>openWindow()</code> multiple times, is left to the user agent implementers to decide on.
```
It should probably throw.

marcoscaceres replied:
(invalid state error)

tommythorsen replied:
Ok. I'll change this.

marcoscaceres commented on index.html:
```diff
@@ -964,9 +970,15 @@
+    The exact behavior in the case where a payment handler calls <code>openWindow()</code> multiple times, is left to the user agent implementers to decide on.
+    User agents MAY, in this case, let subsequent calls to <code>openWindow()</code> resolve to the same <code>WindowClient</code> instance.</p>
+
+  <p>User agents SHOULD display the origin of the payment handler to the user.</p>
```
Should be in the privacy and security considerations section

marcoscaceres commented:
This is not aimed at @tommythorsen - but in general, can we please be extremely disciplined with adding things to this spec. 

We've had to waste a lot of time on the Payment Request spec dealing with technical debt built up by adding bad markup, random phrasing, conformance statements that don't actually apply to any products, etc. 

Can we also be disciplined with running [HTML5 Tidy](https://github.com/htacg/tidy-html5) over the document before any commit goes in - and make sure ReSpec is being utilized to the fullest (i.e., don't use `<code>` tags). And make sure we don't pollute the commit history with random commit messages. 

Also, no one should be committing directly to `gh-pages`. @sideshowbarker, can you please set up branch protection to stop people doing that - i.e., only allow pull requests? 

ReSpec is already screaming that this document has over 66 issues. That's already really alarming - we need to do better and not rush things into the spec. 

If Editors need a "how to best use ReSpec" tutorial during the F2F, I'm happy to run that.

sideshowbarker commented:
> no one should be committing directly to `gh-pages`. @sideshowbarker, can you please set up branch protection to stop people doing that - i.e., only allow pull requests?

I doesn’t seem appropriate for me or anybody else from the W3C staff to try to unilaterally impose branch protection on a repo. That should be something for the people actually doing the work to discuss and decide on. https://github.com/orgs/w3c/teams/payment-app-api-editors shows a number of people with admin access to the repo who can set pull-request-required branch protection for the default branch if that’s what they want.

marcoscaceres commented:
@sideshowbarker, fair. Can someone with admin please put branch protection on?

tommythorsen commented:
@marcoscaceres 
> Can someone with admin please put branch protection on?

Done!

adrianhopebailie commented:
> This is not aimed at @tommythorsen - but in general, can we please be extremely disciplined with adding things to this spec.

> We've had to waste a lot of time on the Payment Request spec dealing with technical debt built up by adding bad markup, random phrasing, conformance statements that don't actually apply to any products, etc.

> Can we also be disciplined with running HTML5 Tidy over the document before any commit goes in - and make sure ReSpec is being utilized to the fullest (i.e., don't use <code> tags). And make sure we don't pollute the commit history with random commit messages.

@marcoscaceres - thanks for this comment. I think I speak for all of the editors (@ianbjacobs , @adamroach , @tommythorsen , @jnormore and myself) when I say that we're fairly new to this (and if not then fairly new to specs that include WebIDL and browser algorithms or new to ReSpec or GitHub or all of it).

So, I am 100% in agreement that we try and avoid accumulating debt that we need to deal with later but also appeal to you to help us out here.

Feel free to point out things we're doing wrong (as you have above) or share some pro tips because we're keen to do them right.

While I appreciate you are putting a lot of effort into the PR spec right now your help as an editor here would be appreciated if you can spare some time. I think that at a minimum @tommythorsen , you and I could have an occasional call to sync up since timezones make it hard for you guys to join the TF calls.

Commit:
Add an algorithm for opening a window

tommythorsen commented:
I think I have addressed all of @marcoscaceres' review comments. I have introduced a few more respec warnings, but I intend to take a closer look at the whole list of respec issues in another PR, after this one is finished.

marcoscaceres commented:
Thanks for the update. Will follow up soon. 

> On 16 Mar 2017, at 8:37 pm, Tommy Thorsen <notifications@github.com> wrote:
> 
> I think I have addressed all of @marcoscaceres' review comments. I have introduced a few more respec warnings, but I intend to take a closer look at the whole list of respec issues in another PR, after this one is finished.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ianbjacobs commented:
Thank you @tommythorsen! I have looked at the changes. I look forward to @marcoscaceres' review. 

@jakearchibald, your input would also be appreciated.

Ian

Commit:
fixed up markup a bit

marcoscaceres commented:
A general note on style and invoking algorithms: see how it's done in the Payment Request spec. Ideally, this spec should end up being edited in the same way as that one. 

For other examples of excellent specs whose style we should follow:

 * https://fetch.spec.whatwg.org
 * https://url.spec.whatwg.org

marcoscaceres commented on index.html:
```diff
@@ -843,70 +844,100 @@
+          <table>
+            <tr>
+              <th>Internal Slot</th>
+              <th>Description (<em>non-normative</em>)</th>
```
If we are going to have defaults, we should have a default column in the table.

marcoscaceres commented on index.html:
```diff
@@ -843,70 +844,100 @@
+        <section>
+          <h3>Handling a payment request event</h3>
+          <p>Upon receiving a <a>payment request</a> by way of
+          <code>PaymentRequest.show()</code> and subsequent user
```
Nit: `<code><a data-cite="!payment-request#show-method">PaymentRequest.show()</a></code>`

marcoscaceres commented on index.html:
```diff
@@ -843,70 +844,100 @@
+          <p>Upon receiving a <a>payment request</a> by way of
+          <code>PaymentRequest.show()</code> and subsequent user
+          selection of a payment handler, the <a>user agent</a>
+          MUST run the following steps or their equivalent:</p>
```
nit: remove "their equivalent"

marcoscaceres commented on index.html:
```diff
@@ -843,70 +844,100 @@
+          MUST run the following steps or their equivalent:</p>
+          <ol>
+            <li>Let <var>registration</var> be the <a>service
+            worker registration</a> corresponding to the payment
```
Expected link to "payment handler" definition

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+      than one client window using this method.</p>
+      <section>
+        <h3><dfn>Open Window Algorithm</dfn></h3>
+        <p class="note">This algorithm is <i>very</i> similar to
```
nit: class="ednote" ... should probably file a new bug for this, then we can link it to the spec.

marcoscaceres replied:
See also: https://github.com/w3c/respec/wiki/Referencing-GitHub-issues-in-your-spec

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+        <h3><dfn>Open Window Algorithm</dfn></h3>
+        <p class="note">This algorithm is <i>very</i> similar to
+        the <a href=
+        "https://www.w3.org/TR/service-workers/#clients-openwindow">
```
Nit: don't link to specs directly. instead:

```
<a data-cite="!service-workers/#clients-openwindow">Open Window Algorithm</a>
```

marcoscaceres replied:
And yes, we should do the security checks, and then defer to the the service worker algorithm.

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+          <li>Let <var>request</var> be the <a>PaymentRequest</a>
+          that triggered this payment request event.
+          </li>
+          <li>If <var>request</var>.[[\state]] is not
```
nit: link to `<a>[[\state]]</a>`

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+          that triggered this payment request event.
+          </li>
+          <li>If <var>request</var>.[[\state]] is not
+          "interactive", then return a Promise rejected with a <a>
```
Nit: link and define "interactive"

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+          <li>If <var>request</var>.[[\state]] is not
+          "interactive", then return a Promise rejected with a <a>
+            DOMException</a> whose value is
+            "<a>InvalidStateError</a>".
```
Please see: https://www.w3.org/2001/tag/doc/promises-guide with regards to phrasing of these. It's close, but could be better.

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+            DOMException</a> whose value is
+            "<a>InvalidStateError</a>".
+          </li>
+          <li>Let <var>url</var> be the result of parsing the
```
Note, this can throw.

marcoscaceres replied:
Also, we should invoke the [[!WHATWG-URL]] parser here.

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+          <li>Let <var>url</var> be the result of parsing the
+          <var>url</var> argument.</li>
+          <li>If <var>url</var> is failure, return a Promise
+          rejected with a <a>TypeError</a>.
```
You should re-throw the error instead of creating an empty one.

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+              <li>Let <var>newContext</var> be a new <a>top level
+              browsing context</a>.
+              </li>
+              <li>Navigate <var>newContext</var> to <var>url</var>,
```
Nit: link to html navigate.

marcoscaceres replied:
Same with the exception types

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+          </li>
+          <li>Let <var>promise</var> be a new <a>Promise</a>.
+          </li>
+          <li>Run these steps in parallel:
```
nit: link to in parallel

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+              <li>If the navigation throws an exception, reject
+              <var>promise</var> with that exception and abort
+              these steps.</li>
+              <li>If the origin is not the same as the
```
Need to clear which origin (the origin of the newContext). Also, what happens if the newContext goes from:

```
newContext->foo.com REDIRECT newContext->bar.com REDIRECT newContext->foo.com

```

So probably need to track navigations.

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+              <var>event</var>.<a>appRequest</a>.<a href=
+              "#dom-paymentapprequest-origin">origin</a>, then:
+                <ol class="algorithm">
+                  <li>Resolve promise with null.</li>
```
Wonder why this doesn't reject with a SecurityError?

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+              <a href=
+              "https://www.w3.org/TR/service-workers/#capture-windowclient-algorithm">
+                Capture Window Client</a> algorithm, or its
+                equivalent, with <var>newContext</var> as the
```
nit: drop "or its equivalent" (yes, I know it's in the SW spec... it's silly there too :))

dlongley commented:
The general approach of opening the window via a call on the event (to make it context-aware) LGTM.

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+              </li>
+              <li>If <var>event</var>.<a>[[\windowClient]]</a> is
+              not null, then:
+                <ol class="algorithm">
```
nit, I think you only need "algorithm" on the top-most `ol`

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+              not null, then:
+                <ol class="algorithm">
+                  <li>If
+                  <var>event</var>.<a>[[\windowClient]]</a>.visibilityState
```
nit: link to .visibilityState, so one can see what other states it could possibly be in.

marcoscaceres commented on index.html:
```diff
@@ -843,70 +844,100 @@
       [Exposed=ServiceWorker]
       interface PaymentRequestEvent : ExtendableEvent {
         readonly attribute PaymentAppRequest appRequest;
+        <a>Promise</a>&lt;<a>WindowClient</a>&gt; openWindow(DOMString url);
```
This has to be a USVString. It can't be a DOMString. You can't pass a DOMString representing a URL.

marcoscaceres replied:
Don't add `a` elements inside IDL. ReSpec will do the linking for you.

marcoscaceres commented on index.html:
```diff
@@ -843,70 +844,100 @@
+            handler selected by the user.
+            </li>
+            <li>If <var>registration</var> is not found, reject the
+            Promise that was created by
```
Nit: promise should be linked or treated as a concept. In either case, it should be linked to something.

marcoscaceres commented on index.html:
```diff
@@ -843,70 +844,100 @@
+            <a>DOMException</a> whose value
+            "<a>InvalidStateError</a>" and terminate these steps.
+            </li>
+            <li>Invoke the <a>Handle Functional Event</a> algorithm
```
Nit: lowercase the algorithm name.

marcoscaceres commented on index.html:
```diff
@@ -843,70 +844,100 @@
+            </li>
+            <li>If <var>registration</var> is not found, reject the
+            Promise that was created by
+            <code>PaymentRequest.show()</code> with a
```
We might want to abstract away ".show()", as in something can request request for payment, but not necessarily the Payment Request API.

marcoscaceres commented on index.html:
```diff
@@ -843,70 +844,100 @@
+            <var>registration</var> and <var>callbackSteps</var>
+            set to the following steps:
+              <ol>
+                <li>Set <var>global</var> to the global object that
```
I'm confused, where was "global" passed?

tommythorsen replied:
I have no idea. Note that my PR does not actually write all of this. I merely put it inside a section and re-indent it, so that it looks a bit like I've written it. We can have a look at this in a different Issue/PR.

marcoscaceres replied:
Sure, np.

marcoscaceres commented on index.html:
```diff
@@ -843,70 +844,100 @@
+              <ol>
+                <li>Set <var>global</var> to the global object that
+                was provided as an argument.</li>
+                <li>Create a <a>trusted event</a>, <var>e</var>,
```
Use WHATWG-DOM fire an event here instead of the following subsets.

tommythorsen replied:
Same as above. We can have a look at this in a different Issue/PR, since I didn't write this.

marcoscaceres commented on index.html:
```diff
@@ -843,70 +844,100 @@
+                which does not bubble, cannot be canceled, and has
+                no default action.
+                </li>
+                <li>Set the <code>appRequest</code> attribute of
```
nit: `<code>` should be `<a>`

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
-      the checkout context.</p>
-      <p>User agents SHOULD display the origin of a running payment
-      handler to the user.</p>
+      <p>A payment handler that requires visual display and user
```
Wait? who is the payment handler here? User code or the browser arbitrarily calling some JS function?

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+      interaction, may call <a>openWindow()</a>. Since user agents
+      know that this method is connected to the payment request
+      event, they SHOULD render the window in a way that is
+      consistent with the flow and not confusing to the user.</p>
```
This whole assertion seems very wishy-washy, tbh. Maybe this should be a note?

marcoscaceres commented on index.html:
```diff
@@ -927,16 +958,91 @@
+      know that this method is connected to the payment request
+      event, they SHOULD render the window in a way that is
+      consistent with the flow and not confusing to the user.</p>
+      <p>The resulting window client is bound to the tab/window
```
Same with this... all this reads very "non-normative".

marcoscaceres commented on index.html:
```diff
@@ -634,9 +634,10 @@
-      conveyed using the following dictionary:
+    <section data-dfn-for="PaymentAppRequest" data-link-for="PaymentAppRequest">
+      <h3>Payment App Request</h3>
+      <p>The <a>payment app request</a> is
```
this is redundant. Talk about `PaymentAppRequest` instead.

marcoscaceres commented on index.html:
```diff
@@ -843,70 +844,100 @@
+          associated with this <a>payment request</a>.</p>
+        </section>
+        <section>
+          <h3><dfn data-lt="openWindow()">openWindow</dfn>
```
nit: `<dfn>openWindow()</dfn>`

marcoscaceres commented on index.html:
```diff
@@ -843,70 +844,100 @@
+          <h3><dfn data-lt="openWindow()">openWindow</dfn>
+          method</h3>
+          <p>This method is used by the payment handler to show a
+          window to the user. It invokes the <a>Open Window
```
Maybe say: "When called, it runs the ..." in lower case.

marcoscaceres commented on index.html:
```diff
@@ -843,70 +844,100 @@
+          Algorithm</a>.</p>
+        </section>
+        <section>
+          <h3><dfn>respondWith</dfn> method</h3>
```
nit:  missing "()"

marcoscaceres commented:
> @marcoscaceres - thanks for this comment. I think I speak for all of the editors (@ianbjacobs , @adamroach , @tommythorsen , @jnormore and myself) when I say that we're fairly new to this (and if not then fairly new to specs that include WebIDL and browser algorithms or new to ReSpec or GitHub or all of it).

Good guides if getting started:

  * [API Design Principles](https://w3ctag.github.io/design-principles/)
  * [Writing Promise-Using Specifications](http://www.w3.org/2001/tag/doc/promises-guide)
  * [ReSpec](https://github.com/w3c/respec/wiki)
 
>So, I am 100% in agreement that we try and avoid accumulating debt that we need to deal with later but also appeal to you to help us out here.

I'm happy to help. Please tag me on reviews and I'll get to them as quickly as possible. I'll have more time once we put Payment Request into CR.

tommythorsen commented:
@marcoscaceres: Thanks for the pointers!

Did you have any specific change requests? This PR is blocked from merging by your earlier change request, so if you're happy with the current state of the PR, perhaps you would like to change to approval?

marcoscaceres commented:
Odd? I added 30+ change requests earlier today? 

> On 17 Mar 2017, at 8:52 pm, Tommy Thorsen <notifications@github.com> wrote:
> 
> @marcoscaceres: Thanks for the pointers!
> 
> Did you have any specific change requests? This PR is blocked from merging by your earlier change request, so if you're happy with the current state of the PR, perhaps you would like to change to approval?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

marcoscaceres commented:
@tommythorsen see my change requests above. I can see them all there

tommythorsen commented:
30 change requests from today? The only change requests I can see are the 7 outdated ones from two days ago.

tommythorsen commented:
@marcoscaceres: I still see no new change requests from you. I have not received an email about them either. Can you double-check that they have been submitted?

Can anyone else see these new change requests? @ianbjacobs? @adrianhopebailie?

marcoscaceres commented:
Argh, my bad. Seems I didn't hit the request changes button.

Commit:
Adress Marcos' review comments

Commit:
One more tiny change

marcoscaceres reviewed (approved):
So, approving this for merge so we can deal with the much larger issues around event dispatching. All that needs to be rewritten.

marcoscaceres commented:
I feel like we should still pull everything out of this spec an slowly add it back in one part at a time. We've again gone rushed a whole bunch of stuff into a spec without it being carefully checked 😭.
